### DISCUSSION TITLE
v5.3.0-alpha1 - Color mode switch implementation.

### DISCUSSION THREAD

**Discussion by Alex-Toucan** [twbs/bootstrap#37764] [Q&A] [answered]
Hello,
I have [attempted](https://beta.atproducts.xyz/beta/) to add switch off the documentation with the [correct JavaScript](https://github.com/twbs/bootstrap/blob/main/site/static/docs/5.3/assets/js/color-modes.js). Clicking any option doesn't do a thing. What would be the right way to implement this switch? What have I done wrong with this attempt?

**Comment by julien-deramond** [comment, marked as answer]
Hello @Alex-Toucan 

The error in the console is:
```
Uncaught TypeError: btnToActive.querySelector(...) is null
    showActiveTheme https://beta.atproducts.xyz/static/darkmode.js:33
    <anonymous> https://beta.atproducts.xyz/static/darkmode.js:50
    EventListener.handleEvent* https://beta.atproducts.xyz/static/darkmode.js:49
    <anonymous> https://beta.atproducts.xyz/static/darkmode.js:62
```

The HTML markup seems to be OK in the switch dropdown. If I display <code>document.querySelector(`[data-bs-theme-value="light"]`)</code> in the console manually, I can get the button so that's OK.

~~I'd say that the script is loaded before the dropdown is ready/rendered. In our script, everything is triggered by `window.addEventListener('DOMContentLoaded', () => {`. In your case, it is maybe too soon. You could try to replace it by something like `window.addEventListener('load', (event) => {` or `window.onload = (event) => {`.~~

**Edit with the correct answer:**

> The error comes from `btnToActive.querySelector('svg use')`. In your website, you don't use the same structure as in our doc. We use a SVG (`<svg>`), and you're using an icon (`<i>`). So the query selector will return null and break the execution of the script. So you'll need to adapt our script if you want it to work in your case with these icons.

**Reply by Alex-Toucan** [reply]
I have replaced the `window.addEventListener('DOMContentLoaded', () => {` with both of your suggestions. However, both failed.

`window.addEventListener('load', (event) => {`'s console:
```
Uncaught TypeError: Cannot read properties of null (reading 'getAttribute')
    at showActiveTheme (darkmode.js:33:64)
    at darkmode.js:50:5
```

For the `window.onload = (event) => {` suggestion, it removed the auto function, making it default light. Seems like it would be fixable.
```
Uncaught SyntaxError: Unexpected token ')'
```

**Reply by julien-deramond** [reply]
IDK if it'll work but you have an extra `)` right now

```
window.onload = (event) => {
}
```

and not


```
window.onload = (event) => {
})
```

**Reply by Alex-Toucan** [reply]
Looking at sources, seems like this was the issue

Edit: Yea, it was. The auto function is back, but switching still doesn't work.
```
window.onload = (event) => {
    showActiveTheme(getPreferredTheme())

    document.querySelectorAll('[data-bs-theme-value]')
      .forEach(toggle => {
        toggle.addEventListener('click', () => {
          const theme = toggle.getAttribute('data-bs-theme-value')
          localStorage.setItem('theme', theme)
          setTheme(theme)
          showActiveTheme(theme)
        })
      })
  }**)**
})()
```

**Reply by Alex-Toucan** [reply]
Still a bit confused on why switching doesn't work.

**Reply by julien-deramond** [reply]
Sorry I lead you to the wrong path because I was not reading the error correctly. The error comes from `btnToActive.querySelector('svg use')`. In your website, you don't use the same structure as in our doc. We use a SVG (`<svg>`), and you're using an icon (`<i>`). So the query selector will return null and break the execution of the script. So you'll need to adapt our script if you want it to work in your case with these icons.

For a first test, I'd say you can remove the two calls to `showActiveTheme()`. You'll probably have a working theme switch but the active theme won't be shown properly in your dropdown.

**Reply by Alex-Toucan** [reply]
I don't know how the documentation managed to even use the `#sun-fill` for example, as a valid icon. As it would redirect to a "id".
```
<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi me-2 opacity-50 theme-icon bi-sun-fill" viewBox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></svg>
```

**Reply by Alex-Toucan** [reply]
Also, removing the `showActiveTheme()` didn't work as well. Not even the console posted anything.

```
/*!
 * Color mode toggler for Bootstrap's docs (https://getbootstrap.com/)
 * Copyright 2011-2022 The Bootstrap Authors
 * Licensed under the Creative Commons Attribution 3.0 Unported License.
 */

(() => {
  'use strict'

  const storedTheme = localStorage.getItem('theme')

  const getPreferredTheme = () => {
    if (storedTheme) {
      return storedTheme
    }

    return window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light'
  }

  const setTheme = function (theme) {
    if (theme === 'auto' && window.matchMedia('(prefers-color-scheme: dark)').matches) {
      document.documentElement.setAttribute('data-bs-theme', 'dark')
    } else {
      document.documentElement.setAttribute('data-bs-theme', theme)
    }
  }

  setTheme(getPreferredTheme())

  const showActiveTheme = theme => {
    const activeThemeIcon = document.querySelector('.theme-icon-active use')
    const btnToActive = document.querySelector(`[data-bs-theme-value="${theme}"]`)
    const svgOfActiveBtn = btnToActive.querySelector('svg use').getAttribute('href')

    document.querySelectorAll('[data-bs-theme-value]').forEach(element => {
      element.classList.remove('active')
    })

    btnToActive.classList.add('active')
    activeThemeIcon.setAttribute('href', svgOfActiveBtn)
  }

  window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', () => {
    if (storedTheme !== 'light' || storedTheme !== 'dark') {
      setTheme(getPreferredTheme())
    }
  })
})()
```

**Reply by julien-deramond** [reply]
When I said to remove the 2 calls of `showActiveTheme()` I meant rather this:

```js
window.addEventListener('DOMContentLoaded', () => {
    // showActiveTheme(getPreferredTheme())

    document.querySelectorAll('[data-bs-theme-value]')
      .forEach(toggle => {
        toggle.addEventListener('click', () => {
          const theme = toggle.getAttribute('data-bs-theme-value')
          localStorage.setItem('theme', theme)
          setTheme(theme)
          // showActiveTheme(theme)
        })
      })
  })
```

Just to avoid to call this function that won't work in your case and that you'll need to rewrite so that it works with your `<i>`.

**Reply by Alex-Toucan** [reply]
Rewriting for `<i>`. What should I put in after the query selector in the following code?

`const svgOfActiveBtn = btnToActive.querySelector('i').getAttribute('')`

For example, `href` wouldn't quite work.
`<i class="bi bi-twitter abicon"></i>`

**Reply by julien-deramond** [reply]
Quickly tried with your current implementation. Gathered your HTML markup inside Bootstrap doc and the following works since your don't use `href`s with your SVGs:

```js
const showActiveTheme = theme => {
    const activeThemeIcon = document.querySelector('.theme-icon-active use')
    const btnToActive = document.querySelector(`[data-bs-theme-value="${theme}"]`)
    // const svgOfActiveBtn = btnToActive.querySelector('svg use').getAttribute('href')

    document.querySelectorAll('[data-bs-theme-value]').forEach(element => {
      element.classList.remove('active')
    })

    btnToActive.classList.add('active')
    // activeThemeIcon.setAttribute('href', svgOfActiveBtn)
  }
```

It's a bit complicated to help with the code always changing in https://beta.atproducts.xyz/beta/ and not having access to the full code. The most efficient way would be that you create a reproducible example with [CodePen](https://codepen.io) for example. Otherwise I won't be able to find more time on this topic.

**Reply by Alex-Toucan** [reply]
Works! Will make a codepen later today.

**Reply by Alex-Toucan** [reply]
[Here](https://codepen.io/alex-toucan/pen/MWBegPL). The JavaScript doesn't work however and the console doesn't help.

**Reply by julien-deramond** [reply]
Hey. Here is the [modified CodePen](https://codepen.io/julien-deramond/pen/OJwXMLG?editors=1111) where I only commented the call to the cookie consent function:

```js
window.onload = function() { /*cookieConsent();*/ };
```

Switch between light and dark mode works well. I'll let you dig into it but this is not Bootstrap-related.

**Reply by Alex-Toucan** [reply]
[This](https://beta.atproducts.xyz/beta/) is the result. Wouldn't say it's final since I might change my mind on things, but it looks and does great.
